### ISSUE TITLE
Extend SMV so it can run both Scala and Python Modules

### ISSUE THREAD

**Issue by Kai-Chen** [TresAmigosSD/SMV#393]
## Overview

There are 2 ways to start an SMV application
- smv-run - batch run of modules, stages, will launch via PySpark
- smv-shell - interactive shell capable of both Python and Scala mode

The interactive shell can switch between Python and Scala mode by using a special command `%python` and `%scala`.

Design and implementation of `smv-shell` will get its own ticket.
## Main Entry

The main entry always starts with PySpark.  Most of the SMV functionalities are provided in Scala, including Module Resolution (which includes a Module Registry, Loading and Persistence of modules), DataFrame and Column enhancers, Aggregate functions (including UDAFs), DQMs, EDDs, and others.  The Python side will have hashing for the Python modules and will delegate all other functionalities to the Scala implementation.
## Module Hash

Hashing of a Module should depend on both its code and the hash of its parent classes, excluding SmvModule and language base classes.  Hashing of a Scala Module may also depend on its Configuration trait, if present.

If a Python module merely delegates to an underlying implementation in Scala -- as in the case of SmvCsvFile -- it should delegate also the module hash computation.
## Module Dependencies

Modules written in different languages can depend on one another, as long as they are in the same stage.
## Python-Scala Protocol

The application-level communication protocol is yet undetermined.  Two possible candidates are a simple Protocol object and JSON.  As before, we will decide which one is better through implementation and use.

Last part of #337

**Comment by ninjapapa** [comment]
☝️

**Comment by Kai-Chen** [comment]
The py4j library has a `callback_server` which we could use to have a Python implementation of a Java interface, which would represent the Python-Scala interface.  The`callback_server` is not turned on by default in PySpark.  It is used by the `StreamingContext`, however, so there is some example of this in the Spark source tree.

@AliTajeldin has emphasized that we should make sure that we could configure a specific port for the `callback_server` to avoid complications from possible deployment environment and also to allow multiple users launch `smv-pyshell` on the same machine (e.g. a client gateway).

**Comment by Kai-Chen** [comment]
At this stage, we can run *pure* applications of both types (i.e. 100% Scala or 100% Python) from the same commandline `smv-pyrun`.  Next we will tackle the use cases where

- [ ] a Scala module depends on a Python module
- [ ] a Python module depends on a Scala module

Once that's done, all the remains is to migrate the current code  to the right place (circular dependency check, module hashing, DQM, etc)

**Comment by Kai-Chen** [comment]
Module hashing needs to be done before inter-language dependencies can work.  This is because the upstream module may need to persist its result for the downstream module to use (via the `i` runParams), and persisting requires computing the module's HashOfHash.  The stub for the "remote" (cross-language) module does not know how to compute the remote hash; and it would make the module overcomplicated to build into it the knowledge of how to locate its real module.  Hence, the `DataSetRepository` interface needs to have module hashing as part of its API.

**Comment by Kai-Chen** [comment]
- [ ] DQM
